### PR TITLE
bpo-34579: Modify test_embed DEFAULT_CONFIG for AIX

### PR DESCRIPTION
**Pull request by aixtools**
https://bugs.python.org/issue34579

### CODE REVIEW

**Review by vstinner — CHANGES_REQUESTED** [review]
Please don't duplicate the whole DEFAULT_CONFIG, it would be annoying to maintain these two copies.

IHMO the change is wrong. Why would AIX behave differently than other operating systems?

**Comment by aixtools** [comment]
Will be a few days before I have time to make the change.

As to why AIX is different from other platforms:
a) no clue why other platforms print "(null)" when there is nothing.
b) I have not been able to locate the precise code that generates this output. If you can point me to it maybe I can figure out why they differ.

Thanks for the review.

**Comment by vstinner** [comment]
>  a) no clue why other platforms print "(null)" when there is nothing.

You cannot come with a solution if you didn't identify the root issue.

**Comment by vstinner** [comment]
> b) I have not been able to locate the precise code that generates this output. If you can point me to it maybe I can figure out why they differ.

See Programs/_testembed.c. You can run the program directly.

**Comment by aixtools** [comment]
It seems there is a difference is the way AIX libc deals with printing NULL:

root@x066:[/data/prj/python]cat nullpr.c
#include <stdio.h>
main()
{
        char *str = NULL;

        printf("NULL prints as:>>>%s<<<\n", str);
}

root@x066:[/data/prj/python]uname
AIX

root@x066:[/data/prj/python]./nullpr
NULL prints as:>>><<<

root@x074:/data/prj/python# uname
Linux
root@x074:/data/prj/python# ./nullpr
NULL prints as:>>>(null)<<<
root@x074:/data/prj/python#

**Comment by aixtools** [comment]
Note: there is a failed test, but not in anything I have changed:

root@x066:[/data/prj/python/git/python3-3.8]git branch
  bpo-11192
* bpo-34579
  bpo-pr
  master
root@x066:[/data/prj/python/git/python3-3.8]git diff master Lib/test/test_socket.py
root@x066:[/data/prj/python/git/python3-3.8]

I have made the requested changes; please review again.

**Review by vstinner — CHANGES_REQUESTED** [review]
I would prefer to see a constant NULL_STR which would be equal to "(null)" on Linux and "" on AIX.

**Comment by aixtools** [comment]
excellent idea. Wish it had been mine!

I have made the requested changes; please review again.

File: Lib/test/test_embed.py [hunk 1]
```diff
@@ -7,7 +7,9 @@
 import re
 import subprocess
 import sys
-
+import platform
```
Comment (vstinner) [inline Lib/test/test_embed.py]: Please put the import in the sorted order of other imports. And add 1 or 2 lines between imports and the constant.

File: Misc/NEWS.d/next/Tests/2018-09-04-15-16-42.bpo-34579.bp4HdM.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Add AIX specific DEFAULT_CONFIG
```
Comment (vstinner) [inline Misc/NEWS.d/next/Tests/2018-09-04-15-16-42.bpo-34579.bp4HdM.rst]: I would prefer to say: Fix test_embed on AIX.

Maybe put your full name, it's up to you. But we usually write "Patch by xxx."

File: Lib/test/test_embed.py (hunk 1 above)
Comment (aixtools) [reply Lib/test/test_embed.py]: Done. Thanks for your time.

**Comment by aixtools** [comment]
I have made the requested changes; please review again.

**Comment by vstinner** [comment]
Oh no. I was editing the commiting the commit message on my phone when my phone decided that it should be merged with the incomplete commit message :-(

**Pull request merged**